### PR TITLE
[LFXV2-1476] feat(fga): add granular committee member access — roster_viewer and email_viewer

### PR DESCRIPTION
**Pull request by andrest50**
## What changed

This PR splits committee member data access into two distinct scopes, aligning the API with the FGA model changes in `lfx-v2-helm`:

- **`roster_viewer`** — can see member names, roles, voting status, and org info
- **`email_viewer`** — can see email addresses (gated separately)

Previously `GET /committees/{uid}/members/{member_uid}` returned everything including email, and was checked against a single `viewer` relation. Now email lives behind its own endpoint and its own FGA relation.

---

### `GET /committees/{uid}/members/{member_uid}` → checked against `roster_viewer`

Returns profile and role data — no email:

```json
{
  "uid": "...",
  "committee_uid": "...",
  "committee_name": "...",
  "committee_category": "...",
  "username": "...",
  "first_name": "...",
  "last_name": "...",
  "job_title": "...",
  "linkedin_profile": "...",
  "appointed_by": "...",
  "status": "...",
  "role": { "name": "...", "start_date": "...", "end_date": "..." },
  "voting": { "status": "...", "start_date": "...", "end_date": "..." },
  "organization": { "id": "...", "name": "...", "website": "..." },
  "created_at": "...",
  "updated_at": "..."
}
```

### `GET /committees/{uid}/members/{member_uid}/contact` → checked against `email_viewer` _(new)_

Returns only the identifiers needed to look up contact info:

```json
{
  "uid": "...",
  "committee_uid": "...",
  "email": "user@example.com"
}
```

---

## Per-committee member visibility (`member_visibility`)

A new `member_visibility` setting on `CommitteeSettings` controls whether committee members can see each other, without affecting what staff/auditors can see:

| Value | Effect |
|---|---|
| `hidden` (default) | Members cannot see each other — neither `roster_viewer` nor `email_viewer` is set |
| `basic_profile` | Members can see each other's names and roles — sets `committee_for_member_roster_access` to the committee itself |
| `full_profile` | Members can see each other's names, roles, and emails — sets both `committee_for_member_roster_access` and `committee_for_member_email_access` to the committee itself |

These are self-referential FGA references (analogous to `vote_for_participant_result_access`): when set, the FGA model grants all `member`s of the committee the corresponding viewer relation on each other.

---

## OpenSearch indexing

A new index subject `lfx.index.committee_member_sensitive` has been added alongside the existing `lfx.index.committee_member`. This allows the indexer to store email in a separate document type gated by the `email_viewer` relation, keeping sensitive fields out of the general member index.

On member deletion, a delete event is now published to **both** index subjects so that the email document is removed from the sensitive index and does not remain accessible.

---

## Ruleset changes

```yaml
# Before
- id: "rule:lfx:lfx-v2-committee-service:committee_members:get"
  # ...
  relation: viewer

# After — two rules, two relations
- id: "rule:lfx:lfx-v2-committee-service:committee_members:get"
  # ...
  relation: roster_viewer

- id: "rule:lfx:lfx-v2-committee-service:committee_members:get_contact"
  match:
    routes:
      - path: /committees/:uid/members/:member_uid/contact
  # ...
  relation: email_viewer
```

---

## Ticket

[LFXV2-1476](https://linuxfoundation.atlassian.net/browse/LFXV2-1476)

🤖 Generated with [Claude Code](https://claude.ai/code)

[LFXV2-1476]: https://linuxfoundation.atlassian.net/browse/LFXV2-1476?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/service/committee_member_writer.go [hunk 1]
```diff
@@ -790,14 +790,14 @@ func (uc *committeeWriterOrchestrator) publishMemberMessages(ctx context.Context
...
+			AccessCheckRelation:  constants.RelationRosterViewer,
 			HistoryCheckObject:   fmt.Sprintf("committee:%s", data.Member.CommitteeUID),
 			HistoryCheckRelation: "auditor",
 			SortName:             data.Member.FirstName,
 			NameAndAliases:       nameAndAliases,
 			ParentRefs:           []string{fmt.Sprintf("committee:%s", data.Member.CommitteeUID)},
 			Tags:                 data.Member.Tags(),
-			Fulltext:             fmt.Sprintf("%s %s %s %s", data.Member.FirstName, data.Member.LastName, data.Member.Email, data.Member.Organization.Name),
+			Fulltext:             fmt.Sprintf("%s %s %s", data.Member.FirstName, data.Member.LastName, data.Member.Organization.Name),
 		}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/committee_member_writer.go:801]: Already addressed —  is assigned  (not ), so the email field is stripped from the roster index payload. Email is only included in the separate sensitive payload published to `lfx.index.committee_member_sensitive` with `AccessCheckRelation: email_viewer`.

File: cmd/committee-api/service/committee_service.go [hunk 2]
```diff
@@ -283,6 +283,24 @@ func (s *committeeServicesrvc) GetCommitteeMember(ctx context.Context, p *commit
...
+		"member_uid", p.MemberUID,
+	)
+
+	// Execute use case — reuse GetMember; we only expose the email from the result
+	committeeMember, _, err := s.committeeReaderOrchestrator.GetMember(ctx, p.UID, p.MemberUID)
+	if err != nil {
+		return nil, wrapError(ctx, err)
+	}
+
+	return s.convertMemberDomainToContactResponse(committeeMember), nil
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/committee-api/service/committee_service.go:301]: Tests for `GetCommitteeMemberContact` are already in place at `cmd/committee-api/service/committee_service_test.go` (see `TestGetCommitteeMemberContact`).

File: internal/service/committee_member_writer.go [hunk 3]
```diff
@@ -790,14 +790,14 @@ func (uc *committeeWriterOrchestrator) publishMemberMessages(ctx context.Context
...
 		memberData = data.Member
 	case model.ActionDeleted:
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/committee_member_writer.go:804]: Already addressed — the roster index payload is built from `data.Member.CommitteeMemberBase`, which contains no email field. Email is only emitted via the separate sensitive index (`lfx.index.committee_member_sensitive`) with `AccessCheckRelation: email_viewer`.

File: cmd/committee-api/design/type.go [hunk 4]
```diff
@@ -719,7 +746,7 @@ func OrganizationIDAttribute() {
 // MemberVisibilityAttribute is the DSL attribute for the member visibility setting
 func MemberVisibilityAttribute() {
 	dsl.Attribute("member_visibility", dsl.String, "Dertermines the visibility level of members profiles to other members of the same committee", func() {
-		dsl.Enum("hidden", "basic_profile")
+		dsl.Enum("hidden", "basic_profile", "full_profile")
 		dsl.Default("hidden")
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/committee-api/design/type.go:750]: Fixed in ae8db23 — "Dertermines" → "Determines" and added the missing possessive apostrophe in "members' profiles".

File: internal/service/committee_member_writer.go [hunk 5]
```diff
@@ -814,6 +815,39 @@ func (uc *committeeWriterOrchestrator) publishMemberMessages(ctx context.Context
 		return errs.NewUnexpected("failed to build member indexer message", errBuildIndexerMessage)
 	}
 
+	// Build sensitive indexer message (email only), gated by email_viewer.
+	// This is only published for create/update operations since email is not present on delete.
+	var sensitiveIndexerMessageBuild *model.CommitteeIndexerMessage
+	if action == model.ActionCreated || action == model.ActionUpdated {
+		sensitiveIndexerMessage := model.CommitteeIndexerMessage{
+			Action: action,
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/committee_member_writer.go:829]: Fixed in ae8db23 — the sensitive indexer block is now a `switch` that also handles `ActionDeleted`, publishing a delete event to `lfx.index.committee_member_sensitive` so the email document is removed from the index on member deletion.

File: docs/fga-contract.md [hunk 6]
```diff
@@ -61,14 +61,20 @@ These fields are carried inside the message `data` object.
...
+> `roster_viewer: ["*"]` makes the committee's member list (names & roles) publicly readable on public committees. On private committees this relation is omitted and access is inherited via `member from committee_for_member_roster_access` for members, or `auditor` for staff.
 
 #### References
 
 | Reference | Value | Condition |
 |---|---|---|
 | `project` | `CommitteeBase.ProjectUID` | Always |
+| `committee_for_member_roster_access` | `committee:{CommitteeBase.UID}` (self) | Only when `CommitteeSettings.MemberVisibility == "basic_profile"` |
+
+> `committee_for_member_roster_access` is a self-referential pointer. When set to the committee itself, the FGA model grants all `member`s of the committee `roster_viewer` and `email_viewer` access to each other — enabling members to see names, roles, and email addresses of fellow members. When `MemberVisibility` is `"hidden"` (default), this reference is omitted and members cannot see each other.
```
[the remark this answers is not in the document]
Comment (andrest50) [reply docs/fga-contract.md]: Fixed in ae8db23 — updated the References table and prose to document `committee_for_member_roster_access` and `committee_for_member_email_access` as two separate self-referential pointers with their own conditions (`basic_profile`/`full_profile` for roster; `full_profile` only for email). The `hidden` case is also explicitly called out.

File: cmd/committee-api/design/type.go [hunk 7]
```diff
@@ -719,7 +746,7 @@ func OrganizationIDAttribute() {
 // MemberVisibilityAttribute is the DSL attribute for the member visibility setting
 func MemberVisibilityAttribute() {
 	dsl.Attribute("member_visibility", dsl.String, "Dertermines the visibility level of members profiles to other members of the same committee", func() {
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/committee-api/design/type.go]: Fixed in ae8db23 — applied the suggested correction ("Determines") plus the missing possessive apostrophe.

File: internal/service/committee_member_writer.go [hunk 8]
```diff
@@ -814,6 +815,63 @@ func (uc *committeeWriterOrchestrator) publishMemberMessages(ctx context.Context
...
+				NameAndAliases:       []string{data.Member.Email},
+				Fulltext:             data.Member.Email,
+				Tags: []string{
+					data.Member.UID,
+					fmt.Sprintf("committee_member_uid:%s", data.Member.UID),
+					fmt.Sprintf("committee_uid:%s", data.Member.CommitteeUID),
+					fmt.Sprintf("email:%s", data.Member.Email),
+				},
+				ParentRefs: []string{fmt.Sprintf("committee:%s", data.Member.CommitteeUID)},
+			},
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/committee_member_writer.go:842]: Fixed in 7bd0348. Extracted the tag slice into a `sensitiveMemberTags` variable and assigned it to both `CommitteeIndexerMessage.Tags` and `IndexingConfig.Tags`, so both fields are populated from the same source.

File: cmd/committee-api/service/committee_service_test.go [hunk 9]
```diff
@@ -1518,3 +1551,121 @@ func TestLeaveCommittee(t *testing.T) {
...
+		setupMock      func(*mockCommitteeReaderOrchestrator)
+		expectError    bool
+		validateResult func(*testing.T, *committeeservice.CommitteeMemberContactWithReadonlyAttributes)
+		validateCalls  func(*testing.T, []getMemberCall)
+	}{
+		{
+			name: "returns email from CommitteeMemberSensitive",
+			payload: &committeeservice.GetCommitteeMemberContactPayload{
+				UID:       "committee-123",
+				MemberUID: "member-456",
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/committee-api/service/committee_service_test.go:1571]: Fixed in 7bd0348. Added `Version: "1"` to all three `GetCommitteeMemberContactPayload` instances in `TestGetCommitteeMemberContact`.

File: cmd/committee-api/design/type.go [hunk 10]
```diff
@@ -437,26 +438,52 @@ func CommitteeMemberBaseAttributes() {
...
 
 	CommitteeMemberUIDAttribute()
 	CommitteeUIDMemberAttribute()
 	CommitteeNameMemberAttribute()
 	CommitteeCategoryMemberAttribute()
-	CommitteeMemberBaseAttributes()
+	CommitteeMemberBasicBaseAttributes()
 	CreatedAtAttribute()
 	UpdatedAtAttribute()
 })
```
[the remark this answers is not in the document]
Comment (andrest50) [reply cmd/committee-api/design/type.go:475]: It should be fine to not return the email of the committee member on the POST/PUT and other related endpoints. If the email is needed it can be retrieved from the new /contact endpoint.

**Comment by dealako** [comment]
Hey @andrest50, thanks for pushing this through several rounds already. This is a solid piece of work: splitting roster vs. email access into `roster_viewer`/`email_viewer` FGA relations, adding the `member_visibility` setting with a fail-secure `hidden` default, gating the new `/contact` endpoint correctly, and routing email into its own `lfx.index.committee_member_sensitive` subject so it's no longer embedded in the general roster index. I traced the earlier bot-flagged concern about email leaking into the roster indexer payload, and that's been fixed — the roster message builder now excludes email and a separate sensitive message carries it, gated by `email_viewer`.

Three parallel reviews covered security/privacy, correctness/performance/tests, and style/API-compliance/docs. Style, goa API compliance, and documentation (FGA + indexer contracts, migration README) all came back clean.

**Issue count**

- 🔴 Blocking: 0
- 🟡 Minor: 4: (1) migration script exits without flushing buffered NATS publishes on the failure path, risking silently-lost "processed" records; (2) migration script doesn't guard against empty `member.Email`, producing malformed index fields; (3) no test coverage for the migration script; (4) roster and sensitive-email index writes aren't atomic, so a partial publish failure can leave the sensitive (email) index stale with no reconciliation path
- ⚪ Nit: 0
- ❔ Question: 0

No Data Privacy findings survived review — the email-gating boundary itself looks sound; the remaining concerns are about migration robustness and cross-index consistency under partial failure, not exposure.

**Merge conflicts**: this PR currently shows as `CONFLICTING` against `main` and will need a rebase/merge before it can land.

🔴 **Needs changes before approval** — due to the merge conflicts against `main` and the 4 minor findings above (all in the migration script and index-write consistency, not in the FGA/access-control logic itself).

File: scripts/migrations/split_member_index/main.go [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+	slog.InfoContext(ctx, "split_member_index complete",
+		"total", len(memberKeys),
+		"processed", processed,
+		"failed", failed,
+		"dry_run", *dryRun,
+	)
+
+	if failed > 0 {
+		os.Exit(1)
```
Comment (dealako) [inline scripts/migrations/split_member_index/main.go:196]: [minor] Migration exits without flushing buffered NATS publishes on partial failure

Issue: `nc.Publish()` (lines 155, 166) is buffered/async — it queues data client-side and does not guarantee delivery. When `failed > 0`, the script calls `os.Exit(1)` (line 196), which skips the `defer nc.Close()` at line 76 entirely (Go does not run deferred functions on `os.Exit`).
Proof: `nats.Conn.Close()` normally flushes the outbound buffer as part of teardown, so the happy-path exit is fine, but the `os.Exit(1)` branch bypasses that defer, so any messages published in the run's final batch that haven't hit the wire yet can be silently dropped even though they were logged as `processed`.
Why it matters: An operator seeing `failed: N > 0` already knows to investigate/rerun, but they have no way to know that some of the `processed` count was also lost. Since the script is safely rerunnable (it's keyed off KV, not a queue it drains), this is recoverable but worth closing.
Fix: call `nc.Flush()` (or `nc.FlushWithContext(ctx)`) explicitly right before every exit path, or restructure to flush once at end-of-loop before evaluating `failed`.

File: scripts/migrations/split_member_index/main.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+		Fulltext:             fmt.Sprintf("%s %s %s", member.FirstName, member.LastName, member.Organization.Name),
+	}
+
+	return json.Marshal(built)
+}
+
+// buildSensitiveMessage builds the email-only indexer message gated by email_viewer.
+// Payload is {uid, committee_uid, email}. Mirrors the sensitive block in
+// committee_member_writer.go ActionUpdated.
+func buildSensitiveMessage(ctx context.Context, member *model.CommitteeMember) ([]byte, error) {
```
Comment (dealako) [inline scripts/migrations/split_member_index/main.go:257]: [minor] buildSensitiveMessage doesn't guard against empty member.Email

Issue: `buildSensitiveMessage` unconditionally sets `SortName: member.Email`, `NameAndAliases: []string{member.Email}`, and `Fulltext: member.Email` (lines 292-294), and builds an `email:%s` tag from it.
Proof: if a legacy KV record predates strict email validation and has an empty `Email`, this emits an index document with `SortName: ""`, `NameAndAliases: [""]`, and an `email:` tag with no value — the live writer path is gated by validation before a member can be created, but this migration reads existing data directly with no such guard.
Why it matters: malformed index documents (empty sort key, empty-string alias) can produce odd search/sort behavior for those records.
Fix: skip or warn-and-continue when `member.Email == ""` rather than emitting a document built from an empty string.

File: scripts/migrations/split_member_index/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
+# split_member_index
```
Comment (dealako) [inline scripts/migrations/split_member_index/README.md:1]: [minor] No test coverage for the migration script

Issue: `scripts/migrations/split_member_index/` has no `_test.go` files.
Proof: the script's build/publish logic (`buildBothMessages`, `buildRosterMessage`, `buildSensitiveMessage`) is unexported and untested; a one-time backfill with no test coverage for its message-building or partial-failure handling is a common source of silent data issues once run against production KV data.
Why it matters: combined with the flush and empty-email gaps above, there's no automated check that this script does what the README says it does before someone runs it against prod.
Fix: add unit tests for the three build functions (including an empty-email case) and, if feasible, an integration test against a fake KV bucket verifying both subjects get published per member.

File: internal/service/committee_member_writer.go [hunk 14]
```diff
@@ -847,6 +907,12 @@ func (uc *committeeWriterOrchestrator) publishMemberMessages(ctx context.Context
 		func() error {
 			return uc.committeePublisher.Indexer(ctx, constants.IndexCommitteeMemberSubject, indexerMessageBuild, sync)
```
Comment (dealako) [inline internal/service/committee_member_writer.go:908]: [minor] Roster and sensitive-email index writes are not atomic; no reconciliation on partial failure

Issue: the roster index publish (line 908) and the sensitive email index publish (line 914) run as independent tasks in the same worker pool (`concurrent.NewWorkerPool(len(messages)).Run`, line 948).
Proof: the worker pool is errgroup-based and returns the first error, cancelling remaining not-yet-started work. If the roster publish for an `ActionUpdated` (e.g., an email correction) succeeds but the sensitive publish fails, the roster index reflects the new state while `lfx.index.committee_member_sensitive` — the sole enforcement boundary for `email_viewer` — keeps the stale email indefinitely. There's no reconciliation/backfill sweep in the codebase (only the one-off `scripts/migrations/split_member_index`) that would detect or repair this drift.
Why it matters: a stale sensitive-index document containing an old email that should have been corrected can persist silently, which is exactly the kind of data-quality gap the dedicated sensitive index is supposed to avoid.
Fix: either couple the two publishes so a single failure aborts both (e.g., publish sensitive first and only proceed to roster on success), or add a periodic reconciliation job comparing roster/sensitive document versions per member and republishing on drift.

**Review by dealako — CHANGES_REQUESTED** [review]
Consolidated review — see the summary comment for full detail: https://github.com/linuxfoundation/lfx-v2-committee-service/pull/79#issuecomment-5627930319

Requesting changes for: merge conflicts against main, plus 4 minor findings (all in the migration script and cross-index write consistency — the FGA/access-control split itself looks sound).

**Comment by andrest50** [comment]
We can revisit this if needed later. Closing right now since there's many conflicts.
